### PR TITLE
9580 cryptography 2.5 openssl 1.1.1 tls 1.3 fixes

### PR DESCRIPTION
**Pull request by markrwilliams**
## Contributor Checklist:

* [x] There is an associated ticket in Trac. https://twistedmatrix.com/trac/ticket/9580
* [x] The changes pass minimal style checks (see: https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted )
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles )
* [x] I have updated the automated tests.

### CODE REVIEW

**Comment by markrwilliams** [comment]
@alex Are you available to review this?

File: src/twisted/test/test_ssl.py [hunk 1]
```diff
@@ -309,33 +305,26 @@ def getHandleExceptionType(self):
...
-            if platform.getType() == 'win32':
-                return errno.WSAENOTSOCK
-
-        # Otherwise, we expect an error about how we tried to write to a
-        # shutdown connection.  This is terribly implementation-specific.
-        return [('SSL routines', 'SSL_write', 'protocol is shutdown')]
+        # We expect an error about how we tried to write to a shutdown
+        # connection.  This is terribly implementation-specific.
+        return hamcrest.contains(
+            hamcrest.contains(
```
Comment (alex) [inline src/twisted/test/test_ssl.py:319]: Is it correct to nest two `contains`? I don't really know things about `hamcrest`

File: src/twisted/test/test_tcp.py [hunk 2]
```diff
@@ -1190,10 +1194,13 @@ def clientDisconnected(result):
             if not server.lostConnectionReason.check(error.ConnectionClosed):
                 err(server.lostConnectionReason,
                     "Server lost connection for unexpected reason")
-            expectedErrorCode = self.getHandleErrorCode()
+            errorCodeMatcher = self.getHandleErrorCodeMatcher()
```
Comment (alex) [inline src/twisted/test/test_tcp.py:1197]: Why is this change necessary?

**Review by alex — COMMENTED** [review]
I feel like the newsfragment should say that twisted now officially supports TLS1.3?

File: src/twisted/test/test_tcp.py (hunk 2 above)
Comment (markrwilliams) [reply src/twisted/test/test_tcp.py:1197]: The change in `test_ssl` that adds the nested `contains` matcher is in a test that inherits from _this_ test, which expects the `getHandleErrorCodeMatcher` "abstract method" to return something to compare against the raised exception's first argument.

Now the OpenSSL implementation detail could be one of two messages, and I happen to have introduced the PyHamcrest dependency, so I decided to address that ambiguity with a matcher, and changed the method name (which is in a test and not `proto_helpers`, so a private API) to reflect that.

**Comment by markrwilliams** [comment]
> I feel like the newsfragment should say that twisted now officially supports TLS1.3?

I don't know that it actually does.  What I know is that it supports enough of TLS 1.3 to make our tests pass!

I guess the point of TLS 1.3 is that supporting its defaults is equivalent supporting TLS 1.3?

**Review by alex — COMMENTED** [review]
This probably is fine, but to be honest I find hamcrest so difficult to read that I'm uncomfortable actually approving this and would suggest a different reviewer should have a look.

File: src/twisted/conch/client/knownhosts.py [hunk 3]
```diff
@@ -489,7 +489,7 @@ def promptResponse(response):
 
                 keytype = key.type()
 
-                if keytype is "EC":
+                if keytype == "EC":
```
Comment (exarkun) [inline src/twisted/conch/client/knownhosts.py:492]: Great.

File: src/twisted/internet/_sslverify.py [hunk 4]
```diff
@@ -1470,7 +1470,7 @@ def __init__(self,
         @type raiseMinimumTo: L{TLSVersion} constant
 
         @param insecurelyLowerMinimumTo: The minimum TLS version to use,
-            possibky lower than Twisted's default.  If not specified, it is a
+            possibly lower than Twisted's default.  If not specified, it is a
```
Comment (exarkun) [inline src/twisted/internet/_sslverify.py:1473]: Nice.

File: src/twisted/internet/_sslverify.py [hunk 5]
```diff
@@ -1764,6 +1764,12 @@ def _expandCipherString(cipherString, method, options):
     try:
         ctx.set_cipher_list(cipherString.encode('ascii'))
     except SSL.Error as e:
+        # OpenSSL 1.1.1 turns an invalid cipher list into TLS 1.3
+        # ciphers, so pyOpenSSL >= 19.0.0 raises an artificial Error
+        # that lacks a corresponding OpenSSL error if the cipher list
+        # consists only of these after a call to set_cipher_list.
+        if not e.args[0]:
+            return []
```
Comment (exarkun) [inline src/twisted/internet/_sslverify.py:1772]: Boy OpenSSL is baaaaaaaaaaaaaaaaaaaaaaad.  But, uh, I guess this one is at least half pyOpenSSL's fault.  Raising a bare `Error`?  pyOpenSSL goes out of its way to identify this case and then encodes that information in this extremely lossy form!  Lame!  I suggest this is worthy of a pyOpenSSL bug report.  The least pyOpenSSL could do is raise an exception specifically encoding the information it has uncovered here.

File: src/twisted/test/test_ssl.py [hunk 6]
```diff
@@ -564,9 +553,6 @@ def checkResult(failure):
             [clientProtocol.done.addErrback(checkResult),
              serverProtocol.done.addErrback(checkResult)])
 
-    if newTLS is None:
-        test_bothSidesLoseConnection.skip = "Old SSL code doesn't always close cleanly."
-
```
Comment (exarkun) [inline src/twisted/test/test_ssl.py:569]: This seems unrelated?

File: src/twisted/test/test_tcp.py [hunk 7]
```diff
@@ -1135,17 +1138,18 @@ def getHandleExceptionType(self):
...
+        socket handle.
         """
         # Windows and Python 3: returns WSAENOTSOCK
         # Windows and Python 2: returns EBADF
         # Linux, FreeBSD, macOS: returns EBADF
         if platform.isWindows() and _PY3:
-            return errno.WSAENOTSOCK
-        return errno.EBADF
+            return hamcrest.equal_to(errno.WSAENOTSOCK)
+        return hamcrest.equal_to(errno.EBADF)
```
Comment (exarkun) [inline src/twisted/test/test_tcp.py:1152]: I might say that it is more in the spirit of matchers for the conditional here to be _inside_ the matcher.  That does involve writing more code, though, and more code is worse than less code, so ... :man_shrugging:

**Review by exarkun — APPROVED** [review]
Thanks.  Looks pretty good.  Some inline comments.  Address as you like and then merge.

File: src/twisted/internet/_sslverify.py (hunk 5 above)
Comment (glyph) [reply src/twisted/internet/_sslverify.py:1772]: Filed: https://github.com/pyca/pyopenssl/issues/819

**Comment by jMyles** [comment]
> I feel like the newsfragment should say that twisted now officially supports TLS1.3?

@alex: Do I understand correctly, though, that there is no `TLSv1_2_METHOD` yet in pyOpenSSL?  Does that mean that, for the moment, even though there's support in Twisted and Cryptography 2.5, that there's still not quite a way to use it in eg `DefaultOpenSSLContextFactory`?  

Or is there another way to do TLS 1.3 that I'm missing?

**Comment by glyph** [comment]
@jMyles Did you mean `TLSv1_3_METHOD`?  Generally `SSLv23_METHOD` is the only method you should use, because it doesn't mean "SSL 2 or SSL 3" it means "negotiate the version of TLS to use based on the handshake".  You can disable individual protocol versions with `SSL_OP_NO_*` as options, rather than methods; or you can use `SSL_CTX_set_min_proto_version`.  See https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_new.html

**Comment by jMyles** [comment]
@glyph: Indeed I did mean `TLSv1_3_METHOD` - that was a typo.

And my use case involves both the client and the server running my code; how can I guarantee the use of TLS 1.3?

**Comment by glyph** [comment]
Sigh. We need support in PyDoctor for our various enum-alikes (and real enums as well) since the answer *should* be on this page:

https://twistedmatrix.com/documents/19.2.1/api/twisted.internet.ssl.TLSVersion.html

**Comment by glyph** [comment]
Here's the official place to find the answer:

https://twisted.readthedocs.io/en/twisted-19.2.0/core/howto/ssl.html#tls-protocol-options

**Comment by glyph** [comment]
In brief: `options = CertificateOptions(..., raiseMinimumTo=TLSVersion.TLSv1_3)`

This is presuming that when TLSv1.4 comes out, you don't want the code to *preclude* negotiating TLS 1.4.
